dtoh: Emit floating point literals according to their base type ...

MoonlightSentinel opened the pull request:
... and respect `NaN` / `Infinity`. The latter is implemented using the `NAN` and `INFINITY` macros from `math.h` to ensure compatibility with < C++ 11.

kinke commented on src/dmd/dtoh.d:
```diff
@@ -2029,8 +2030,30 @@ public:
+            default:
+                printf("%s = %d\n", e.type.toChars(), e.type.ty);
+                assert(false);
+        }
```
Thanks! - Please use the `dmd.root.ctfloat.CTFloat.{isNaN,isInfinity,sprint}` functions to check and print `real_t` values (not necessarily corresponding to the host C `long double`, not just for non-DMD compilers, but also for DMD linked against MSVC). I'd also prefer an uppercase `L` as `long double` suffix. And detecting `-inf` might be required as well.

kinke replied:
The printing logic, incl. falling back to a hex string if the compact decimal representation isn't accurate enough, for non-Nan/inf is already in [hdrgen](https://github.com/dlang/dmd/blob/0470e5988f29fc278cb5cb86a3e42300a8b4d9ce/src/dmd/hdrgen.d#L1883) and could be refactored.

ibuclaw replied:
Agreed, gdc doesn't use native floating point (either?), so casting to float/double is futile. ;-)

ibuclaw replied:
Agreed, gdc doesn't use native floating point (either?), so casting to float/double is futile. ;-)

MoonlightSentinel replied:
Thanks! Changed the implementation to use `CTFloat` and `floatToBuffer`.

MoonlightSentinel committed:
Refactor floatToBuffer into a standalone function

To make it accessible outside of hdrgen.d

MoonlightSentinel committed:
dtoh: Emit floating point literals according to their base type ...

... and respect `NaN` / `Infinity`. The latter is implemented using
the `NAN` and `INFINITY` macros from `math.h` to ensure compatibility
with < C++ 11.
